HACKING-URL.md

```diff
@@ -0,0 +1,319 @@
+The Developer’s Guide to implementing URL handlers in the MusicBrainz Server
+============================================================================
+
+This file describes how to implement specific handlers for external websites in
+MusicBrainz Server to: allow linking to other databases and lyrics website,
+make it easier to add links, prevent editor from entering wrong links,
+improve the display of links on MusicBrainz website.
+
+For easy starting, look at similar
+[resolved MBS "URL cleanup" tickets](https://tickets.metabrainz.org/issues/?jql=project%20%3D%20MBS%20AND%20component%20%3D%20%22URL%20cleanup%22%20AND%20status%20%3D%20Closed%20AND%20resolution%20IN%20%28Done%2CFixed%29)
+which have comments linking to pull requests. More particularly,
+look at potential STYLE tickets linked to these MBS tickets,
+and at the commit messages in the pull requests. Do the same.
+If any doubt, get back to this guide which should contain the answer.
+
+Our URL handling has two main parts: editing handlers (unwanted URL block, URL
+cleanup, relationship type autoselection, relationship validation), and
+display handlers (with favicons, in relationship lists, in the sidebar).
+
+Table of Contents
+-----------------
+
+<!-- toc -->
+
+- [Prerequisites](#prerequisites)
+- [URL editing handlers](#url-editing-handlers)
+  * [Cleanup](#cleanup)
+  * [Autoselection](#autoselection)
+  * [Validation](#validation)
+  * [Supporting new relationship types](#supporting-new-relationship-types)
+  * [Entity-wide rules](#entity-wide-rules)
+- [URL display handlers](#url-display-handlers)
+  * [Favicons](#favicons)
+  * [Sidebar display](#sidebar-display)
+  * [Rendered scheme override](#rendered-scheme-override)
+  * [Custom in-page display](#custom-in-page-display)
+  * [Miscellaneous display customization](miscellaneous-display-customization)
+
+<!-- tocstop -->
+
+Prerequisites
+-------------
+
+Allowing to link external websites as
+either [other databases](https://musicbrainz.org/doc/Other_Databases_Relationship_Type/Whitelist) (for persistence issues)
+or [lyrics websites](https://musicbrainz.org/doc/Style/Relationships/URLs/Lyrics_whitelist) (for license issues)
+must be approved by a STYLE decision beforehand.
+See <https://musicbrainz.org/doc/Proposals>
+for how to request such STYLE decision,
+before even thinking about coding handlers.
+
+Other relationship types don’t require any STYLE decision,
+but having a self-assigned MBS ticket is still required.
+
+The following research work is usually expected about external websites to be handled:
+
+* Form and number of its URLs in the MusicBrainz database already;
+  - Tip: For a first approximate count, you can query the MusicBrainz API about URLs with (approximate) patterns;
+  - For example [`url:*/openlibrary.org/*`](https://musicbrainz.org/ws/2/url?query=url:*/openlibrary.org/*&limit=1&fmt=json) counts Open Library URLs (mainly).
```

Comment:
It is not intended to be “formal”, just concise, especially not to lengthen prerequisites too much. Can we add these details when we get to the section about editing handlers?

Semi-colons in lists are very common. See [https://en.wikipedia.org/wiki/Punctuation](https://en.wikipedia.org/wiki/Punctuation) for example.

Reply:
We don't get charged by the word, why not be explanatory and clear rather than concise? We want a new person reading this to understand why we do things and have an easy way to do them.